Add terraform fmt command after templates are rendered.

mjhuber opened the pull request:
implements #27 
After templates are rendered, runs `terraform fmt` in the destination directory.  Exits gracefully if an error occurred.

sudermanjr commented:
Would it also be good to run a `terraform validate` at the same time?

mjhuber commented:
> Would it also be good to run a `terraform validate` at the same time?

oooh, excellent idea!  I'll add that in.

ejether commented on pentagon_datadog/monitor.py:
```diff
@@ -39,6 +40,15 @@ def add(self, destination, overwrite=False):
+                logging.debug("terraform fmt output:\n{}".format(tf))
+
+                validate = subprocess.check_output(['terraform', 'validate', '--check-variables=false', destination])
+            except subprocess.CalledProcessError as validateErr:
```
What is the case where `subprocess.CalledProcessError` will not be raised and the general `Exception` will?

mjhuber replied:
when i was testing I couldn't get `terraform fmt` to throw `CalledProcessError`.  I added the general exception so it would never throw an unhandled exception for formatting.

ejether replied:
Interesting, thats is odd. I'm getting it to throw in a smoke test. Do you think the double `exception` is still necessary?

mjhuber replied:
you got it to throw `CalledProcessError`?

ejether replied:
I did get a smoke test to throw that error:
```
import subprocess
subprocess.check_output(['terraform', 'fmt', './'])
```
 but I don't _know_ that it is going to behave the same way in context of the application.

sudermanjr commented:
Tests pass for me locally, but they are failing in CI for some reason

ejether commented:
I'll bet terraform is not installed on the images running the test.

sudermanjr commented:
Also, a note, the terraform fmt has changed the test files when I run them.  We should probably check-in formatted versions of the files so that local testing doesn't cause changes in the repo

mjhuber commented:
> I'll bet terraform is not installed on the images running the test.

you're right - no terraform.  I'll work on fixing that. 

```
$ whereis terraform
terraform:
```

ejether commented:
In the interests of keeping the `dashboard.py` and `monitor.py` DRY, would it make sense to move the `fmt` and related call to `rodd.py` (or elsewhere) and make a call to it instead of repeating the code?

mjhuber commented:
> The terraform fmt has changed the test files when I run the tests. We should probably check-in formatted versions of the files so that local testing doesn't cause changes in the repo

I added the formatted test files and realized dashboards weren't getting the terraform love, so I added terraform fmt / validate to the dashboard class too.

mjhuber commented:
> In the interests of keeping the `dashboard.py` and `monitor.py` DRY, would it make sense to move the `fmt` and related call to `rodd.py` (or elsewhere) and make a call to it instead of repeating the code?

Good idea.  I moved the validate to the base class and incremented the version in setup.py.  I'll include #50 as part of the release.

ejether commented on CHANGELOG.md:
```diff
@@ -4,6 +4,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+=======
```
Whoops. Looks like a rebase added some stuff maybe?

ejether commented on pentagon_datadog/rodd.py:
```diff
@@ -114,6 +115,16 @@ def _create_tf_file(self, data):
             logging.error(e)
             logging.debug(traceback.format_exc(e))
 
+    def _validate_tf(self, destination):
```
Looks great

ejether commented on Makefile:
```diff
@@ -0,0 +1,13 @@
+##
```
I have mixed feelings about adding `Make` into our builds. 
It introduces a new pattern. Do you anticipate adding other tasks as well or see the Make adding value beyond this case?

sudermanjr replied:
Same hesitation here.  This would be a new pattern in any of our projects.  I'm not wholly against it, but it would be new.

sudermanjr commented on CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@ All notable changes to this project will be documented in this file.
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+- after templates are rendered, `terraform fmt` is run.
```
Should this be updated to include the `validate` as well?

mjhuber replied:
👍 yeah, added to the changelog.

I also added a check for terraform files in the dir before running the validate commands, which eliminates any `Error: No Terraform configuration files found in directory` when no monitors are defined in a rodd config yaml.
